osbuild.py

```diff
@@ -196,6 +193,60 @@ def print_header(title, options, machine_name):
     print(f"\t# nsenter -a --wd=/root -t `machinectl show {machine_name} -p Leader --value`")
     print()
 
+class Stage:
+    def __init__(self, stage, objects):
+        m = hashlib.sha256()
+        m.update(json.dumps(stage, sort_keys=True).encode())
+
+        self.name = stage["name"]
+        self.options = stage.get("options", {})
+        self.id = m.hexdigest()
+        self.base = stage.get("base")
+        self.resources = stage.get("systemResourcesFromEtc", [])
+        self.objects = objects
+
+    def run(self, interactive=False):
+        output_tree = os.path.join(self.objects, self.id)
+
+        if os.path.exists(output_tree):
+            return
+
+        with BuildRoot() as buildroot, tmpfs() as tree:
+            if interactive:
+                print_header(f"{self.name} ({self.id})", self.options, buildroot.machine_name)
+            if self.base:
+                input_tree = os.path.join(self.objects, self.base)
+                subprocess.run(["cp", "-a", f"{input_tree}/.", tree], check=True)
+            r = buildroot.run_stage(self.name, self.options, self.resources, tree, interactive)
+            shutil.rmtree(output_tree, ignore_errors=True)
+            os.makedirs(output_tree, mode=0o755)
+            subprocess.run(["cp", "-a", f"{tree}/.", output_tree], check=True)
```

Comment:
I agree that this is super useful, but I think saving every stage needs to be optional unless we have something like ostree as a backing store. Possibly even turned off by default.

Composer needs a way to save at specific points (e.g., after a base tree has been assembled). Can we do that with this model?

Reply:
Hm, I was thinking of that as a future optimization. I mean, the model now is that everything is implicitly stored, and there is no way to get access to a base, other than one that was previously stored (there is no notion of one being before the other in the pipeline or anything like that). Derisive as I was at first, maybe it would make sense in the end to use ostree as a backing store, rather than the current `cp -a`.

We may be able to come up with some way to not store all the intermediary objects, but do you agree to postone that for later as an optimization?